Validator Registration/Unregistration

flotothemoon opened the pull request:
Adds support for Validator registration/unregistration, implementing the first part of the Validator FSM. Specifically, this PR adds two new particles, `UnregisteredValidatorParticle` and `RegisteredValidatorParticle`, both carrying the address of the validator in question and a nonce. The first unregistered Validator particle for an address (with nonce 0) is virtualised as having an `UP` spin to initialise the FSM. 

The higher-level virtualisation of unregistered Validator particles requires support in the lower-level primitives of `SysCall` and `ConstraintScryptEnv`. Additionally, constraint scripts were (marginally) cleaned up by the introduction of a builder for ParticleDefinitions, reducing the complexity of overloaded sys calls.

Commit:
Add registered and unregistered Particles

Commit:
Expose generalized virtualization support

In prep for Validator FSM (where unregistered needs to be virtualised UP)

Commit:
Implement ValidatorConstraintScrypt

First pass of Validator constraints implementation, describing transitions for validator FSM (registered <=> unregistered). Also virtualise unregistered validator particles as UP, adding a nonce to ensure uniqueness.

Commit:
Introduce Builder for ParticleDefinition

Commit:
Use ParticleDefinition only instead of overloaded sys calls

flotothemoon commented on src/main/java/com/radixdlt/engine/RadixEngine.java:
```diff
@@ -118,6 +118,9 @@ public void store(Atom atom) throws RadixEngineException {
 		synchronized (stateUpdateEngineLock) {
 			stateCheckAndStore(atom, cmAtom);
+
+			// TODO update validator state here
```
Left for a future PR.

talekhinezh replied:
RE doesn't know about "validators". Comment should say something like "TODO return updated state for some given query"

flotothemoon replied:
Ah, true that, will fix

Commit:
Add documentation for new ParticleDefinition and ValidatorConstraintScrypt

Commit:
Add tests for static checks of ValidatorConstraintScrypt

Commit:
Move transition procedure implementation into inner class for easier testing

Commit:
Add ValidatorConstraintScrypt test suite

Commit:
Add ParticleDefinition.Builder.build documentation

Commit:
Add tests for ValidatorConstraintScrypt unregistered validator virtualisation

As per the Validator FSM, unregistered validators with nonce zero should be virtualised as being in the UP state.

clstrfsck commented on examples/src/main/java/com/radixdlt/examples/tictactoe/TicTacToeBaseParticle.java:
```diff
@@ -47,6 +51,8 @@
 	@DsonOutput(Output.ALL)
 	private final ImmutableList<TicTacToeSquareValue> board;
 
+	private final Supplier<Set<RadixAddress>> playersSupplier = Suppliers.memoize(() -> ImmutableSet.of(getXPlayer(), getOPlayer()));
```
This seems overcomplicated for a demo app.  Maybe better just to construct a fresh set in getPlayers(), or in the constructor?

flotothemoon replied:
Didn't realise this was part of the demo when introducing this, changing back. Good point

clstrfsck commented on src/main/java/com/radixdlt/atommodel/tokens/TokensConstraintScrypt.java:
```diff
@@ -35,32 +36,32 @@
-			particle -> particle.getRRI().getAddress(),
-			TokenDefinitionUtils::staticCheck,
-			MutableSupplyTokenDefinitionParticle::getRRI
+		os.registerParticle(MutableSupplyTokenDefinitionParticle.class, ParticleDefinition.<MutableSupplyTokenDefinitionParticle>builder()
```
I don't really like the formatting here, as it the chained methods from the builder are at the wrong indent.  Prefer the previous approach
```
object.methodName(
    firstArgOfSeveral,
    secondArgOfSeveral
        .chainedMethod1(arg, arg)
        .chainedMethod2(etc, andSoOn)
);
```

clstrfsck replied:
This applies to all the methods here.

flotothemoon replied:
Ah yes, sure.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/ValidatorConstraintScrypt.java:
```diff
@@ -0,0 +1,163 @@
+			// ensure nonce is increased when transitioning between the two states
+			long inputNonce = inputNonceMapper.applyAsLong(inputParticle);
+			long outputNonce = outputNonceMapper.applyAsLong(outputParticle);
+			if (inputNonce + 1 != outputNonce) {
```
Noice.  Was looking for this, and was not disappointed :)

flotothemoon replied:
:)

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/ValidatorConstraintScrypt.java:
```diff
@@ -0,0 +1,163 @@
+			long outputNonce = outputNonceMapper.applyAsLong(outputParticle);
+			if (inputNonce + 1 != outputNonce) {
+				return Result.error(String.format(
+					"output nonce must be input nonce + 1, but %d != %d + 1",
```
You can just use %s (which just calls toString() on the wrapped type) for formatting everything in Java, which makes things easier when moving from primitives to objects.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/ValidatorConstraintScrypt.java:
```diff
@@ -0,0 +1,163 @@
+			UnregisteredValidatorParticle::getNonce);
+	}
+
+	private <I extends Particle, O extends Particle> void createTransition(SysCalls os,
```
This formatting seems ... well, problematic and hard to read.

Can we keep with the current approach:
```
void methodName(
    Type1 arg1,
    Type2 arg2,
) {
    // Code goes here
}
```

flotothemoon replied:
Sure, good point

clstrfsck commented on src/main/java/com/radixdlt/atommodel/unique/UniqueParticleConstraintScrypt.java:
```diff
@@ -17,18 +17,18 @@
 package com.radixdlt.atommodel.unique;
 
+import com.radixdlt.atomos.ParticleDefinition;
 import com.radixdlt.atomos.Result;
```
Result import no longer used.

clstrfsck commented on src/main/java/com/radixdlt/atomos/ConstraintScryptEnv.java:
```diff
@@ -21,6 +21,7 @@
 import com.google.common.reflect.TypeToken;
 import com.radixdlt.atommodel.routines.CreateCombinedTransitionRoutine;
 import com.radixdlt.constraintmachine.Particle;
+import com.radixdlt.constraintmachine.Spin;
```
Doesn't seem to be used.

clstrfsck commented on src/main/java/com/radixdlt/atomos/ConstraintScryptEnv.java:
```diff
@@ -34,6 +35,7 @@
 import java.util.Collections;
```
Doesn't seem to be used.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/ValidatorConstraintScrypt.java:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	@VisibleForTesting
+	static class ValidatorTransitionProcedure<I extends Particle, O extends Particle> implements TransitionProcedure<I, VoidUsedData, O, VoidUsedData> {
```
Checkstyle has this line down as being too long.

flotothemoon replied:
Ah, indeed.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/ValidatorConstraintScrypt.java:
```diff
@@ -0,0 +1,163 @@
+		private final Function<O, RadixAddress> outputAddressMapper;
+		private final ToLongFunction<O> outputNonceMapper;
+
+		ValidatorTransitionProcedure(Function<I, RadixAddress> inputAddressMapper,
```
Prefer formatting change as discussed elsewhere.

clstrfsck commented on src/main/java/com/radixdlt/atomos/CMAtomOS.java:
```diff
@@ -31,33 +31,31 @@
-	private static final ParticleDefinition<Particle> VOID_PARTICLE_DEF = new ParticleDefinition<>(
-		v -> {
+	private static final ParticleDefinition<Particle> VOID_PARTICLE_DEF = ParticleDefinition.builder()
+		.addressMapper(v -> {
```
Could just have `v -> throw new UnsupportedOperationException("message")` instead of braces.

flotothemoon replied:
Really? Doesn't compile for me

clstrfsck commented on src/main/java/com/radixdlt/atomos/CMAtomOS.java:
```diff
@@ -31,33 +31,31 @@
-		},
-		v -> {
+		})
+		.staticValidation(v -> {
```
Similarly here.

clstrfsck commented on src/main/java/com/radixdlt/atomos/ParticleDefinition.java:
```diff
@@ -58,7 +63,88 @@
+			}
+
+			// cast as necessary
+			return new ParticleDefinition<>(
```
This is a mess from a type safety point of view, and I appreciate that it's not a mess that you made entirely by yourself.  I really think this should be tidied up.

flotothemoon replied:
Totally agreed. I've added a comment here explaining as much. Unfortunately, fixing this properly would require some more substantial changes, so it will have to do for now.

clstrfsck commented on src/main/java/com/radixdlt/atomos/ParticleDefinition.java:
```diff
@@ -58,7 +63,88 @@
+		private boolean allowsTransitionsFromOutsideScrypts;
+
+		private Builder() {
+		}
```
Ideally throwing an IllegalStateException here.

flotothemoon replied:
Why would I do that? I need to be able to create an instance of `Builder` in `builder()` above. I just don't want people using the constructor directly since it's messier.

clstrfsck replied:
Ah, OK.   My bad.

clstrfsck commented on src/main/java/com/radixdlt/atomos/ParticleDefinition.java:
```diff
@@ -58,7 +63,88 @@
+		private Function<T, Result> staticValidation = x -> Result.success();
+		private Function<T, RRI> rriMapper;
+		private Function<T, Spin> virtualizeSpin;
+		private boolean allowsTransitionsFromOutsideScrypts;
```
Maybe should be an explicit default to `false` here, rather than relying on the default initialisation.

flotothemoon replied:
Sure can do

clstrfsck commented on src/test/java/com/radixdlt/atommodel/validators/ValidatorConstraintScryptTest.java:
```diff
@@ -0,0 +1,216 @@
+package com.radixdlt.atommodel.validators;
```
Missing header comments.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/RegisteredValidatorParticle.java:
```diff
@@ -0,0 +1,41 @@
+package com.radixdlt.atommodel.validators;
```
Missing header comments.

clstrfsck commented on src/test/java/com/radixdlt/atommodel/validators/ValidatorConstraintScryptTest.java:
```diff
@@ -0,0 +1,216 @@
+/**
+ * Exercise the critical invariants of the {@link ValidatorConstraintScrypt}.
+ *
+ * The invariants tested here follow the flow of the validator FSM:
```
Kudos for docs, but list needs to be formatted as Javadoc to be readable in Javadoc output.
Something like:
```
/**
 * Exercise the critical invariants of the {@link ValidatorConstraintScrypt}.
 *
 * The invariants tested here follow the flow of the validator FSM:
 * <ul>
 *   <li>unregistered validator with nonce 0 is virtualised as UP</li>
 *   <li>validator address must be specified</li>
 *   <li>validator transitions must be between validators of the same address</li>
 *   <li>the nonce must increase by exactly one with every transition</li>
 *   <li>validator transitions must be signed by the address</li>
 * </ul>
 *
 * These tests are implemented by either testing the static check directly or the generated transition procedure.
 */
```

clstrfsck commented on src/test/java/com/radixdlt/atommodel/validators/ValidatorConstraintScryptTest.java:
```diff
@@ -0,0 +1,216 @@
+ * These tests are implemented by either testing the static check directly or the generated transition procedure.
+ */
+public class ValidatorConstraintScryptTest {
+	private static Function<Particle, Result> staticCheck;
```
Not sure why these are initialised once for the whole class.  Feels like they should not be static, and the initialisation should be @ Before, not @ BeforeClass.

flotothemoon replied:
Why does it feel like they shouldn't be static?

flotothemoon replied:
They're supposed to be the same for all tests

clstrfsck replied:
I think they should be equivalent for each test, but having them use the exact same objects is potentially problematic, eg having state retained between test methods may result in tests executed earlier having an influence on the outcome of tests executed later.

flotothemoon replied:
Fair enough. Currently, this shouldn't be an issue by design, but if it ever becomes one, this isn't the right place to test it. Will change.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/ValidatorConstraintScrypt.java:
```diff
@@ -0,0 +1,163 @@
+package com.radixdlt.atommodel.validators;
```
Missing header comments.

clstrfsck commented on src/main/java/com/radixdlt/atommodel/validators/UnregisteredValidatorParticle.java:
```diff
@@ -0,0 +1,41 @@
+package com.radixdlt.atommodel.validators;
```
Missing header comments.

clstrfsck commented on src/main/java/com/radixdlt/atomos/ParticleDefinition.java:
```diff
@@ -18,35 +18,40 @@
 	}
 
-	Function<T, Stream<RadixAddress>> getAddressMapper() {
+	Function<T, Set<RadixAddress>> getAddressMapper() {
```
Not really sure why this changed from a Stream to a Set?

flotothemoon replied:
Easier to work with downstream since I need to check the size before accessing for validation

clstrfsck reviewed (changes requested):
Some missing header comments need adding, but also some other comments worth considering.

Commit:
Simplify TicTacToeBaseParticle

Commit:
Fix documentation and code formatting issues

flotothemoon committed:
Merge branch 'infrastructure/generic-atom' into feature/validator-fsm

Commit:
Make ValidatorConstraintScryptTest use separate check instances per test

Commit:
Add test for MessageParticleConstraintScrypt
